<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.251">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2023-08-07">

<title>Contributions to Giotto</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="contribute_files/libs/clipboard/clipboard.min.js"></script>
<script src="contribute_files/libs/quarto-html/quarto.js"></script>
<script src="contribute_files/libs/quarto-html/popper.min.js"></script>
<script src="contribute_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="contribute_files/libs/quarto-html/anchor.min.js"></script>
<link href="contribute_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="contribute_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="contribute_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="contribute_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="contribute_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Contributions to Giotto</h1>
</div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">August 7, 2023</p>
    </div>
  </div>
    
  </div>
  

</header>

<section id="how-to-contribute" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="how-to-contribute"><span class="header-section-number">1</span> How to contribute</h2>
<p>We welcome contributions or suggestions from other developers and here we provide a number of rules, suggestions and tips to help you contribute to Giotto in a sustainable manner. Please contact us if you have questions or would like to discuss an addition or major modifications to the Giotto main code.</p>
<p><span class="math inline">\(~\)</span></p>
<section id="coding-style" class="level3" data-number="1.1">
<h3 data-number="1.1" class="anchored" data-anchor-id="coding-style"><span class="header-section-number">1.1</span> Coding Style</h3>
<p>Following a particular programming style will help programmers read and understand source code conforming to the style, and help to avoid introducing errors. Here we present a small list of guidelines on what is considered a good practice when writing R codes in Giotto package. Most of them are adapted from <a href="https://bioconductor.org/developers/how-to/coding-style/">Bioconductor - coding style</a> or <a href="https://google.github.io/styleguide/Rguide.xml">Google’s R Style Guide</a>. These guidelines are preferences and strongly encouraged!</p>
<ul>
<li><strong>Naming</strong>
<ul>
<li>Use camelCase for function names.</li>
<li>Use snake_case for non-exported functions, like <code>'non_exported_function'</code></li>
<li>Use snake_case for parameter names.<br>
</li>
<li>Do not use “.” as a separator (in the S3 class system, some(x) where x is class A will dispatch to some.A).</li>
</ul></li>
<li><strong>Use of space</strong>
<ul>
<li>Do not place a space before a comma, but always place one after a comma. This: <code>a, b, c</code>.<br>
</li>
<li>Always use space around “=” when using named arguments to functions. This: <code>somefunc(a = 1, b = 2)</code>.</li>
</ul></li>
<li><strong>Use of symbols</strong>
<ul>
<li>Do not use any non-UTF-8 characters unless provided as the unicode symbol. For example: <code>'\u00F6'</code> for <code>ö</code></li>
</ul></li>
</ul>
<p>Beyond these guidelines, <a href="https://github.com/r-lib/styler"><em>styler</em></a> should be used in order to maintain code uniformity.</p>
<p><span class="math inline">\(~\)</span></p>
</section>
<section id="stat-functions" class="level3" data-number="1.2">
<h3 data-number="1.2" class="anchored" data-anchor-id="stat-functions"><span class="header-section-number">1.2</span> Stat functions</h3>
<p>Most Giotto commands can accept several matrix classes (DelayedMatrix, SparseM, Matrix or base matrix). To facilitate this we provide <strong>flexible</strong> wrappers that work on any type of matrix class. They can be found in the <a href="https://github.com/drieslab/Giotto/blob/suite/R/utilities.R">utilities.R</a> file.</p>
<ul>
<li><strong>mean_flex</strong>: analogous to mean()<br>
</li>
<li><strong>rowSums_flex</strong>: analogous to rowSums()<br>
</li>
<li><strong>rowMeans_flex</strong>: analogous to rowMeans()<br>
</li>
<li><strong>colSums_flex</strong>: analogous to colSums()<br>
</li>
<li><strong>colMeans_flex</strong>: analogous to colMeans()<br>
</li>
<li><strong>t_flex</strong>: analogous to t()<br>
</li>
<li><strong>cor_flex</strong>: analogous to cor()</li>
</ul>
<p><span class="math inline">\(~\)</span></p>
</section>
<section id="auxiliary-functions" class="level3" data-number="1.3">
<h3 data-number="1.3" class="anchored" data-anchor-id="auxiliary-functions"><span class="header-section-number">1.3</span> Auxiliary functions</h3>
<p>Giotto has a number of auxiliary or convenience functions that might help you to adapt your code or write new code for Giotto. We encourage you to use these small functions to maintain uniformity throughout the code.</p>
<ul>
<li><strong>lapply_flex</strong>: analogous to lapply() and works for both windows and unix systems<br>
</li>
<li><strong>all_plots_save_function</strong>: compatible with Giotto instructions and helps to automatically save generated plots<br>
</li>
<li><strong>determine_cores</strong>: to determine the number of cores to use if a user does not set this explicitly<br>
</li>
<li><strong>get_os</strong>: to identify the operating system<br>
</li>
<li><strong>update_giotto_params</strong>: will catch and store the parameters for each used command on a giotto object</li>
<li><strong>package_check</strong>: to check if a package exists, works for packages on CRAN, Bioconductor and Github.</li>
</ul>
<p>The last function should be used within your contribution code. It has the additional benefit that it will suggest the user how to download the package if it is not available. To keep the size of Giotto within limits we prefer not to add too many new dependencies.</p>
<p><span class="math inline">\(~\)</span></p>
</section>
<section id="getters-and-setters" class="level3" data-number="1.4">
<h3 data-number="1.4" class="anchored" data-anchor-id="getters-and-setters"><span class="header-section-number">1.4</span> Getters and Setters</h3>
<p>Giotto stores information in different <a href="getting_started_gobject.html#giotto-object-structure">slots</a>, which can be accessed through these getters and setters functions. They can be found in the <a href="https://github.com/drieslab/Giotto/blob/suite/R/accessors.R">accessors.R</a> file.</p>
<ul>
<li><p><strong>get_expression_values</strong>: To select the expression matrix to use</p></li>
<li><p><strong>set_expression_values</strong>: Sets a new expression matrix to the expression slot</p></li>
<li><p><strong>get_spatial_locations</strong>: Get spatial locations to use</p></li>
<li><p><strong>set_spatial_locations</strong>: Sets new spatial locations</p></li>
<li><p><strong>get_dimReduction</strong>: To select the dimension reduction values to use<br>
</p></li>
<li><p><strong>set_dimReduction</strong>: Sets new dimension reduction object</p></li>
<li><p><strong>get_NearestNetwork</strong>: To select the nearest neighbor network (kNN or sNN) to use<br>
</p></li>
<li><p><strong>set_NearestNetwork</strong>: Sets a new nearest neighbor network (kNN or sNN)</p></li>
<li><p><strong>get_spatialNetwork</strong>: To select the spatial network to use<br>
</p></li>
<li><p><strong>set_spatialNetwork</strong>: Sets a new spatial network</p></li>
<li><p><strong>get_spatialGrid</strong>: To select the spatial grid to use</p></li>
<li><p><strong>set_spatialGrid</strong>: Sets a new spatial grid</p></li>
<li><p><strong>get_polygon_info</strong>: Gets spatial polygon information</p></li>
<li><p><strong>set_polygon_info</strong>: Set new spatial polygon information</p></li>
<li><p><strong>get_feature_info</strong>: Gets spatial feature information</p></li>
<li><p><strong>set_feature_info</strong>: Sets new spatial feature information</p></li>
<li><p><strong>get_spatial_enrichment</strong>: Gets spatial enrichment information</p></li>
<li><p><strong>set_spatial_enrichment</strong>: Sets new spatial enrichment information</p></li>
</ul>
<p><span class="math inline">\(~\)</span></p>
</section>
<section id="python-code" class="level3" data-number="1.5">
<h3 data-number="1.5" class="anchored" data-anchor-id="python-code"><span class="header-section-number">1.5</span> Python code</h3>
<p>To use Python code we prefer to create a python wrapper/functions around the python code, which can then be sourced by reticulate. As an example we show the basic principles of how we implemented the Leiden clustering algorithm.</p>
<ol type="1">
<li>write python wrapper and store as python_leiden.py in <em>/inst/python</em>:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> igraph <span class="im">as</span> ig</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> leidenalg <span class="im">as</span> la</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> networkx <span class="im">as</span> nx</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> python_leiden(df, partition_type, initial_membership<span class="op">=</span><span class="va">None</span>, weights<span class="op">=</span><span class="va">None</span>, n_iterations<span class="op">=</span><span class="dv">2</span>, seed<span class="op">=</span><span class="va">None</span>, resolution_parameter <span class="op">=</span> <span class="dv">1</span>):</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># create networkx object</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    Gx <span class="op">=</span> nx.from_pandas_edgelist(df <span class="op">=</span> df, source <span class="op">=</span> <span class="st">'from'</span>, target <span class="op">=</span>  <span class="st">'to'</span>, edge_attr <span class="op">=</span> <span class="st">'weight'</span>)  </span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get weight attribute</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>    myweights <span class="op">=</span> nx.get_edge_attributes(Gx, <span class="st">'weight'</span>)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>    ....</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span>(leiden_dfr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>source python code with reticulate:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>python_leiden_function <span class="ot">=</span> <span class="fu">system.file</span>(<span class="st">"python"</span>, <span class="st">"python_leiden.py"</span>, <span class="at">package =</span> <span class="st">'Giotto'</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>reticulate<span class="sc">::</span><span class="fu">source_python</span>(<span class="at">file =</span> python_leiden_function)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="3" type="1">
<li>use python code as if R code:<br>
See <strong>doLeidenCLuster</strong> for more detailed information.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a> pyth_leid_result <span class="ot">=</span> <span class="fu">python_leiden</span>(<span class="at">df =</span> network_edge_dt,</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">partition_type =</span> partition_type,</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">initial_membership =</span> init_membership,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">weights =</span> <span class="st">'weight'</span>,</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">n_iterations =</span> n_iterations,</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">seed =</span> seed_number,</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">resolution_parameter =</span> resolution)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><span class="math inline">\(~\)</span></p>
</section>
<section id="example" class="level3" data-number="1.6">
<h3 data-number="1.6" class="anchored" data-anchor-id="example"><span class="header-section-number">1.6</span> Example</h3>
<p>As an example we show the implementation of SPARK, which is a recent method developed by <a href="https://doi.org/10.1038/s41592-019-0701-7">Sun et al</a> and provide some comments within the code.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>spark <span class="ot">=</span> <span class="cf">function</span>(gobject,</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">percentage =</span> <span class="fl">0.1</span>,</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">min_count =</span> <span class="dv">10</span>,</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">expression_values =</span> <span class="st">'raw'</span>,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">num_core =</span> <span class="dv">5</span>,</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>                 <span class="at">covariates =</span> <span class="cn">NULL</span>,</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>                 <span class="at">return_object =</span> <span class="st">'data.table'</span>,</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>                 ...) {</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># data.table variables; this is necessary when setting new variables within a data.table</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>  genes <span class="ot">=</span>  adjusted_pvalue <span class="ot">=</span> combined_pvalue <span class="ot">=</span> <span class="cn">NULL</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># test if SPARK is installed </span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># if false, it will suggest how to install SPARK</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">package_check</span>(<span class="at">pkg_name =</span> <span class="st">'SPARK'</span>,</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>                <span class="at">repository =</span> <span class="fu">c</span>(<span class="st">'github'</span>),</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>                <span class="at">github_repo =</span> <span class="st">'xzhoulab/SPARK'</span>)</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># print message with information and encouraging users to cite the authors their work  </span></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">message</span>(<span class="st">"using 'SPARK' for spatial gene/pattern detection. If used in published research, please cite:</span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a><span class="st">  Sun, Shiquan, Jiaqiang Zhu, and Xiang Zhou. “Statistical Analysis of Spatial Expression Pattern for Spatially Resolved Transcriptomic Studies.”</span></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a><span class="st">          BioRxiv, October 21, 2019, 810903. https://doi.org/10.1101/810903."</span>)</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>  <span class="co"># extract expression values from gobject using one of the accessors</span></span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>  expr <span class="ot">=</span> <span class="fu">get_expression_values</span>(<span class="at">gobject =</span> gobject, <span class="at">values =</span> expression_values)</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>  <span class="co"># extract coordinates from gobject</span></span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>  <span class="co"># check the different slots of the giotto S4 object</span></span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>  locs <span class="ot">=</span> <span class="fu">as.data.frame</span>(gobject<span class="sc">@</span>spatial_locs)</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(locs) <span class="ot">=</span> <span class="fu">colnames</span>(expr)</span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Here we implemented spark according to their github example code: </span></span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a>  <span class="co"># create SPARK object for analysis and filter out lowly expressed genes</span></span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a>  sobject <span class="ot">=</span> SPARK<span class="sc">::</span><span class="fu">CreateSPARKObject</span>(<span class="at">counts =</span> expr,</span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">location =</span> locs[,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>],</span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">percentage =</span> percentage,</span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">min_total_counts =</span> min_count)</span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a>  <span class="co"># total counts for each cell</span></span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a>  sobject<span class="sc">@</span>lib_size <span class="ot">=</span> <span class="fu">apply</span>(sobject<span class="sc">@</span>counts, <span class="dv">2</span>, sum)</span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a>  <span class="co"># extract covariates to adjust for from the cell metadata </span></span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="sc">!</span><span class="fu">is.null</span>(covariates)) {</span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a>    <span class="co"># first filter giotto object based on spark object</span></span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a>    filter_cell_ids <span class="ot">=</span> <span class="fu">colnames</span>(sobject<span class="sc">@</span>counts)</span>
<span id="cb4-52"><a href="#cb4-52" aria-hidden="true" tabindex="-1"></a>    filter_gene_ids <span class="ot">=</span> <span class="fu">rownames</span>(sobject<span class="sc">@</span>counts)</span>
<span id="cb4-53"><a href="#cb4-53" aria-hidden="true" tabindex="-1"></a>    tempgobject <span class="ot">=</span> <span class="fu">subsetGiotto</span>(gobject, <span class="at">cell_ids =</span> filter_cell_ids, <span class="at">gene_ids =</span> filter_gene_ids)</span>
<span id="cb4-54"><a href="#cb4-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-55"><a href="#cb4-55" aria-hidden="true" tabindex="-1"></a>    metadata <span class="ot">=</span> <span class="fu">pDataDT</span>(tempgobject)</span>
<span id="cb4-56"><a href="#cb4-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-57"><a href="#cb4-57" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="sc">!</span>covariates <span class="sc">%in%</span> <span class="fu">colnames</span>(metadata)) {</span>
<span id="cb4-58"><a href="#cb4-58" aria-hidden="true" tabindex="-1"></a>      <span class="fu">warning</span>(covariates, <span class="st">' was not found in the cell metadata of the giotto object, will be set to NULL </span><span class="sc">\n</span><span class="st">'</span>)</span>
<span id="cb4-59"><a href="#cb4-59" aria-hidden="true" tabindex="-1"></a>      covariates <span class="ot">=</span> <span class="cn">NULL</span></span>
<span id="cb4-60"><a href="#cb4-60" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb4-61"><a href="#cb4-61" aria-hidden="true" tabindex="-1"></a>      covariates <span class="ot">=</span> metadata[[covariates]]</span>
<span id="cb4-62"><a href="#cb4-62" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb4-63"><a href="#cb4-63" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb4-64"><a href="#cb4-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-65"><a href="#cb4-65" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Fit statistical model under null hypothesis</span></span>
<span id="cb4-66"><a href="#cb4-66" aria-hidden="true" tabindex="-1"></a>  sobject <span class="ot">=</span> SPARK<span class="sc">::</span><span class="fu">spark.vc</span>(sobject,</span>
<span id="cb4-67"><a href="#cb4-67" aria-hidden="true" tabindex="-1"></a>                            <span class="at">covariates =</span> covariates,</span>
<span id="cb4-68"><a href="#cb4-68" aria-hidden="true" tabindex="-1"></a>                            <span class="at">lib_size =</span> sobject<span class="sc">@</span>lib_size,</span>
<span id="cb4-69"><a href="#cb4-69" aria-hidden="true" tabindex="-1"></a>                            <span class="at">num_core =</span> num_core,</span>
<span id="cb4-70"><a href="#cb4-70" aria-hidden="true" tabindex="-1"></a>                            <span class="at">verbose =</span> F,</span>
<span id="cb4-71"><a href="#cb4-71" aria-hidden="true" tabindex="-1"></a>                            ...)</span>
<span id="cb4-72"><a href="#cb4-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-73"><a href="#cb4-73" aria-hidden="true" tabindex="-1"></a>  <span class="co"># test spatially expressed pattern genes</span></span>
<span id="cb4-74"><a href="#cb4-74" aria-hidden="true" tabindex="-1"></a>  <span class="co"># calculating pval</span></span>
<span id="cb4-75"><a href="#cb4-75" aria-hidden="true" tabindex="-1"></a>  sobject <span class="ot">=</span> SPARK<span class="sc">::</span><span class="fu">spark.test</span>(sobject,</span>
<span id="cb4-76"><a href="#cb4-76" aria-hidden="true" tabindex="-1"></a>                              <span class="at">check_positive =</span> T,</span>
<span id="cb4-77"><a href="#cb4-77" aria-hidden="true" tabindex="-1"></a>                              <span class="at">verbose =</span> F)</span>
<span id="cb4-78"><a href="#cb4-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-79"><a href="#cb4-79" aria-hidden="true" tabindex="-1"></a>  <span class="co"># return results</span></span>
<span id="cb4-80"><a href="#cb4-80" aria-hidden="true" tabindex="-1"></a>  <span class="co"># return full output or a simple data.table format with the essental information</span></span>
<span id="cb4-81"><a href="#cb4-81" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(return_object <span class="sc">==</span> <span class="st">'spark'</span>){</span>
<span id="cb4-82"><a href="#cb4-82" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(sobject)</span>
<span id="cb4-83"><a href="#cb4-83" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> <span class="cf">if</span>(return_object <span class="sc">==</span> <span class="st">'data.table'</span>){</span>
<span id="cb4-84"><a href="#cb4-84" aria-hidden="true" tabindex="-1"></a>    DT_results <span class="ot">=</span> data.table<span class="sc">::</span><span class="fu">as.data.table</span>(sobject<span class="sc">@</span>res_mtest)</span>
<span id="cb4-85"><a href="#cb4-85" aria-hidden="true" tabindex="-1"></a>    gene_names <span class="ot">=</span> <span class="fu">rownames</span>(sobject<span class="sc">@</span>counts)</span>
<span id="cb4-86"><a href="#cb4-86" aria-hidden="true" tabindex="-1"></a>    DT_results[, genes <span class="sc">:</span><span class="er">=</span> gene_names]</span>
<span id="cb4-87"><a href="#cb4-87" aria-hidden="true" tabindex="-1"></a>    data.table<span class="sc">::</span><span class="fu">setorder</span>(DT_results, adjusted_pvalue, combined_pvalue)</span>
<span id="cb4-88"><a href="#cb4-88" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(DT_results)</span>
<span id="cb4-89"><a href="#cb4-89" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb4-90"><a href="#cb4-90" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>